### PR TITLE
openmoji-black,openmoji-color: fix hanging builds

### LINKED ISSUE #167869: openmoji has started taking forever to build.
### Describe the bug

openmoji-color and openmoji-black now both take >12h to build. They are timing out on hydra https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.openmoji-color.x86_64-linux.

### Steps To Reproduce
1. nix-build -A openmoji-color

### Notify maintainers
@fgaz

### PR DESCRIPTION
**Pull request by fgaz**
###### Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/167869

ZHF: #172160
@NixOS/nixos-release-managers

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/openmoji/default.nix [hunk 1]
```diff
@@ -14,6 +17,34 @@ let
...
+        sha256 = "u3D9od2xLECNEHhZ+8dkuv9818tPkdP6y/Tvd9CADJg=";
+      })
+      (fetchpatch {
+        url = "https://github.com/fontforge/fontforge/commit/69e263b2aff29ad22f97f13935cfa97a1eabf207.patch";
+        sha256 = "06yyf90605aq6ppfiz83mqkdmnaq5418axp9jgsjyjq78b00xb29";
+      })
+    ];
+    buildInputs = old.buildInputs ++ [ libuninameslist ];
+  });
+  scfbuild-with-pinned-fontforge = scfbuild.override (old: {
```
Comment (kevincox) [inline pkgs/data/fonts/openmoji/default.nix:44]: Nit: It seems weird to call one the date and one just "pinned". Maybe just call both `-pinned`?

**Review by kevincox — APPROVED** [review]

File: pkgs/data/fonts/openmoji/default.nix (hunk 1 above)
Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix]: I prefer using the version then. Fixed.

**Comment by jtojnar** [comment]
I tried running bisection on fontforge itself using [nix-bisect](https://github.com/timokau/nix-bisect) after applying the patch below:

```ShellSession
$ git bisect start 20220308 20201107
$ extra-bisect run \
  nix-build-status \
    --option max-silent-time 240 \
    -f ~/Projects/nixpkgs \
    openmoji-color
```

Unfortunately, the automatic bisection did not work for some reason. Possibly because the [timeout regex](https://github.com/timokau/nix-bisect/blob/3b18985913cf32e62f3a5d3f9ccafb507fed9080/nix_bisect/nix.py#L26) does not match the error message: `error: building of '/nix/store/6z3q2pb6frq47ch617nayaambb7rsyxa-openmoji-13.1.0.drv!out' from .drv file timed out after 240 seconds of silence`

```patch
--- a/pkgs/data/fonts/openmoji/default.nix
+++ b/pkgs/data/fonts/openmoji/default.nix
@@ -3,6 +3,10 @@
 , fetchFromGitHub
 , scfbuild
 , nodejs
+, fontforge
+, libuninameslist
+, fetchpatch
+, patch
 , nodePackages
 , python3Packages
 , variant ? "color" # "color" or "black"
@@ -14,6 +18,30 @@ let
     [ "OpenMoji-Color.ttf" "OpenMoji-Black.ttf" ]
     variant;
 
+    fontforge-bisect = fontforge.overrideAttrs (old: rec {
+      version = "bisect";
+      src = /home/jtojnar/Projects/fontforge;
+      patches = [
+        (fetchpatch {
+          url = "https://salsa.debian.org/fonts-team/fontforge/raw/76bffe6ccf8ab20a0c81476a80a87ad245e2fd1c/debian/patches/0001-add-extra-cmake-install-rules.patch";
+          sha256 = "u3D9od2xLECNEHhZ+8dkuv9818tPkdP6y/Tvd9CADJg=";
+        })
+        # (fetchpatch {
+        #   url = "https://github.com/fontforge/fontforge/commit/69e263b2aff29ad22f97f13935cfa97a1eabf207";
+        #   sha256 = "06yyf90605aq6ppfiz83mqkdmnaq5418axp9jgsjyjq78b00xb29";
+        # })
+      ];
+      prePatch = ''
+        patch() {
+          ${patch}/bin/patch "$@"  || echo 'Failed to apply patch, continuing'
+        }
+      '';
+      buildInputs = old.buildInputs ++ [ libuninameslist ];
+    });
+    scfbuild-with-fontforge-bisect = scfbuild.override (old: {
+      fontforge = fontforge-bisect;
+    });
+
 in stdenv.mkDerivation rec {
   pname = "openmoji";
   version = "13.1.0";
@@ -26,7 +54,7 @@ in stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [
-    scfbuild
+    scfbuild-with-fontforge-bisect
     nodejs
     nodePackages.glob
     nodePackages.lodash
```

File: pkgs/data/fonts/openmoji/default.nix [hunk 2]
```diff
@@ -26,7 +57,7 @@ in stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [
-    scfbuild
+    scfbuild-with-fontforge-20201107
```
Comment (SuperSandro2000) [inline pkgs/data/fonts/openmoji/default.nix:60]:
```suggestion
    scfbuild
```

File: pkgs/data/fonts/openmoji/default.nix [hunk 3]
```diff
@@ -14,6 +17,34 @@ let
...
+      (fetchpatch {
+        url = "https://github.com/fontforge/fontforge/commit/69e263b2aff29ad22f97f13935cfa97a1eabf207.patch";
+        sha256 = "06yyf90605aq6ppfiz83mqkdmnaq5418axp9jgsjyjq78b00xb29";
+      })
+    ];
+    buildInputs = old.buildInputs ++ [ libuninameslist ];
+  });
+  scfbuild-with-fontforge-20201107 = scfbuild.override (old: {
+    fontforge = fontforge-20201107;
+  });
```
Comment (SuperSandro2000) [inline pkgs/data/fonts/openmoji/default.nix:46]:
```suggestion
  scfbuild = scfbuild.override (old: {
    inherit fontforge;
  });
```

File: pkgs/data/fonts/openmoji/default.nix [hunk 4]
```diff
@@ -14,6 +17,34 @@ let
     [ "OpenMoji-Color.ttf" "OpenMoji-Black.ttf" ]
     variant;
 
+  # With newer fontforge the build hangs, see
+  # https://github.com/NixOS/nixpkgs/issues/167869
+  # Patches etc taken from
+  # https://github.com/NixOS/nixpkgs/commit/69da642a5a9bb433138ba1b13c8d56fb5bb6ec05
+  fontforge-20201107 = fontforge.overrideAttrs (old: rec {
```
Comment (SuperSandro2000) [inline pkgs/data/fonts/openmoji/default.nix:24]:
```suggestion
  fontforge = fontforge.overrideAttrs (old: rec {
```

**Review by SuperSandro2000 — COMMENTED** [review]
Please keep the original names so that we don't mix two versions together.

File: pkgs/data/fonts/openmoji/default.nix (hunk 4 above)
Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix:24]: isn't that infinite recursion since in nix `let` is recursive?

Comment (SuperSandro2000) [reply pkgs/data/fonts/openmoji/default.nix:24]: 🤔 could be. I am not sure right now.

Then normally something like the following is used

```suggestion
  fontforge' = fontforge.overrideAttrs (old: rec {
```

Comment (kevincox) [reply pkgs/data/fonts/openmoji/default.nix:24]: If the shadowing doesn't work I would rather stick with meaningful names rather than `'`.

**Comment by risicle** [comment]
`nixpkgs-review` happy, macos 10.15

**Pull request merged**